nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java

```diff
@@ -0,0 +1,235 @@
+                    + "This means the bulletins repository restarted its ids. Restarting querying from the beginning.", new Object[]{currMaxId, lastSentBulletinId});
+            lastSentBulletinId = -1;
+        }
+
+        if (currMaxId == lastSentBulletinId) {
+            getLogger().debug("No events to send due to the current max id being equal to the last id that was sent.");
+            return;
+        }
+
+        final String platform = context.getProperty(PLATFORM).evaluateAttributeExpressions().getValue();
+
+        final Map<String, ?> config = Collections.emptyMap();
+        final JsonBuilderFactory factory = Json.createBuilderFactory(config);
+        final JsonObjectBuilder builder = factory.createObjectBuilder();
+
+        final DateFormat df = new SimpleDateFormat(TIMESTAMP_FORMAT);
+        df.setTimeZone(TimeZone.getTimeZone("Z"));
+
+        final long start = System.nanoTime();
+
+        // Create a JSON array of all the events in the current batch
+        final JsonArrayBuilder arrayBuilder = factory.createArrayBuilder();
+        for (final Bulletin bulletin : bulletins) {
+            if(bulletin.getId() > lastSentBulletinId) {
+                arrayBuilder.add(serialize(factory, builder, bulletin, df, platform, nodeId));
+            }
+        }
+        final JsonArray jsonArray = arrayBuilder.build();
+
+        // Send the JSON document for the current batch
+        try {
+            final Transaction transaction = getClient().createTransaction(TransferDirection.SEND);
+            if (transaction == null) {
+                getLogger().debug("All destination nodes are penalized; will attempt to send data later");
+                return;
+            }
+
+            final Map<String, String> attributes = new HashMap<>();
+            final String transactionId = UUID.randomUUID().toString();
+            attributes.put("reporting.task.transaction.id", transactionId);
+
+            final byte[] data = jsonArray.toString().getBytes(StandardCharsets.UTF_8);
+            transaction.send(data, attributes);
+            transaction.confirm();
+            transaction.complete();
+
+            final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start);
+            getLogger().info("Successfully sent {} Bulletins to destination in {} ms; Transaction ID = {}; First Event ID = {}",
+                    new Object[]{bulletins.size(), transferMillis, transactionId, bulletins.get(0).getId()});
+        } catch (final IOException e) {
+            throw new ProcessException("Failed to send Bulletins to destination due to IOException:" + e.getMessage(), e);
+        }
+
+        // Store the id of the last event so we know where we left off
+        try {
+            StateManager stateManager = context.getStateManager();
+            Map<String, String> newMapOfState = new HashMap<>();
+            newMapOfState.put(LAST_EVENT_ID_KEY, String.valueOf(currMaxId));
```

Comment:
More of a style issue, but I wonder if we should use ```Collections.singletonMap(key, value)``` or do you expect /know already that there may be more values?